domain/relation/state/relation_test.go

```diff
@@ -24,17 +25,491 @@ import (
+	relRequirer := charm.Relation{
+		Name:  "req",
+		Role:  charm.RoleRequirer,
+		Scope: charm.ScopeGlobal,
+	}
+	appUUID1 := s.addApplication(c, "application-1")
+	appUUID2 := s.addApplication(c, "application-2")
+	epUUID1 := s.addApplicationEndpointFromRelation(c, appUUID1, relProvider)
+	epUUID2 := s.addApplicationEndpointFromRelation(c, appUUID2, relRequirer)
+	epUUID3 := s.addApplicationEndpointFromRelation(c, appUUID2, relProvider)
+	epUUID4 := s.addApplicationEndpointFromRelation(c, appUUID1, relRequirer)
+
+	// Act
+	ep1, ep2, err := s.state.AddRelation(context.Background(), relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-1",
+		EndpointName:    "prov",
+	}, relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-2",
+		EndpointName:    "req",
+	})
+	c.Assert(err, jc.ErrorIsNil, gc.Commentf("(Act) unexpected error while inserting the first relation: %s",
+		errors.ErrorStack(err)))
+	ep3, ep4, err := s.state.AddRelation(context.Background(), relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-1",
+		EndpointName:    "req",
+	}, relation.CandidateEndpointIdentifier{
+		ApplicationName: "application-2",
+		EndpointName:    "prov",
+	})
+	c.Assert(err, jc.ErrorIsNil, gc.Commentf("(Act) unexpected error while inserting the second relation: %s",
+		errors.ErrorStack(err)))
+
+	// Assert
+	c.Check(ep1, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-1",
+		Relation:        relProvider,
+	})
+	c.Check(ep2, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-2",
+		Relation:        relRequirer,
+	})
+	c.Check(ep3, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-1",
+		Relation:        relRequirer,
+	})
+	c.Check(ep4, gc.Equals, relation.Endpoint{
+		ApplicationName: "application-2",
+		Relation:        relProvider,
+	})
+	epUUIDsByRelID := make(map[int][]string)
+	err = s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		query := `
+SELECT re.endpoint_uuid, r.relation_id
+FROM relation_endpoint re 
+JOIN relation r  ON re.relation_uuid = r.uuid
+`
+		rows, err := tx.QueryContext(ctx, query)
+		if err != nil {
+			return errors.Capture(err)
+		}
+		defer rows.Close()
+		for rows.Next() {
+			var epUUID string
+			var relID int
+			if err := rows.Scan(&epUUID, &relID); err != nil {
+				return errors.Capture(err)
+			}
+			epUUIDsByRelID[relID] = append(epUUIDsByRelID[relID], epUUID)
+		}
+		return nil
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(epUUIDsByRelID, gc.HasLen, 2)
+	c.Check(epUUIDsByRelID[0], jc.SameContents, []string{epUUID1.String(), epUUID2.String()},
+		gc.Commentf("full map: %v", epUUIDsByRelID))
+	c.Check(epUUIDsByRelID[1], jc.SameContents, []string{epUUID3.String(), epUUID4.String()},
+		gc.Commentf("full map: %v", epUUIDsByRelID))
+
+	// check all relation have a status
+	var statuses []string
+	err = s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
```

Comment:
todo: This test is too long, move the call to `s.TxnRunner().StdTxn` to a new helper method.